art/attacks/evasion/imperceptible_asr/imperceptible_asr.py

```diff
@@ -381,13 +441,33 @@ def _approximate_power_spectral_density_tf(
         # return PSD matrix such that shape is (batch_size, window_size // 2 + 1, frame_length)
         return tf1.transpose(psd_matrix_approximated, [0, 2, 1])
 
-    def _approximate_power_spectral_density_torch(self):
-        """Approximate the power spectral density for a perturbation `perturbation` in PyTorch."""
-        raise NotImplementedError
+    def _approximate_power_spectral_density_torch(
+        self, perturbation: "PTensor", psd_maximum_stabilized: "PTensor"
+    ) -> "PTensor":
+        """
+        Approximate the power spectral density for a perturbation `perturbation` in PyTorch.
+
+        See also `ImperceptibleASR._approximate_power_spectral_density_tf`.
+        """
+        import torch
+
+        # compute short-time Fourier transform (STFT)
+        stft_matrix = torch.stft(
+            perturbation, n_fft=self._window_size, hop_length=self._hop_size, win_length=self._window_size, center=False
+        ).to(self.estimator._device)
+        stft_matrix_abs = torch.sqrt(torch.sum(torch.square(stft_matrix), -1))
+
+        # compute power spectral density (PSD)
+        # note: fixes implementation of Qin et al. by also considering the square root of gain_factor
+        gain_factor = 8.0 / 3.0
+        psd_matrix = gain_factor * torch.square(stft_matrix_abs / (2 * self._window_size))
 
-    def _loss_gradient_masking_threshold_torch(self):
-        """Compute loss gradient of the masking threshold loss in PyTorch."""
-        raise NotImplementedError
+        # approximate normalized psd: psd_matrix_approximated = 10^((96.0 - psd_matrix_max + psd_matrix)/10)
+        psd_matrix_approximated = pow(10.0, 9.6) / torch.unsqueeze(psd_maximum_stabilized, 1) * psd_matrix
+
+        # return PSD matrix such that shape is (batch_size, window_size // 2 + 1, frame_length)
+        # return torch.transpose(psd_matrix_approximated, 1, 2)
```

Comment:
Is this line code or comment?

Reply:
Ups... Line 469 should have been removed :see_no_evil: